clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -296,31 +254,78 @@ public void send(ClientRequest request, long now) {
         doSend(request, false, now);
     }
 
-    private void sendInternalMetadataRequest(MetadataRequest metadataRequest, String nodeConnectionId, long now) {
-        ClientRequest clientRequest = new ClientRequest(nodeConnectionId, now, true,
-                nextRequestHeader(ApiKeys.METADATA), metadataRequest, null);
+    private void sendInternalMetadataRequest(MetadataRequest.Builder builder,
+                                             String nodeConnectionId, long now) {
+        ClientRequest clientRequest = newClientRequest(
+                nodeConnectionId, builder, now, true, null);
         doSend(clientRequest, true, now);
     }
 
-    private void doSend(ClientRequest request, boolean isInternalRequest, long now) {
-        String nodeId = request.destination();
-        if (request.header().apiKey() == ApiKeys.API_VERSIONS.id) {
-            if (!canSendApiVersionsRequest(nodeId))
-                throw new IllegalStateException("Attempt to send API Versions request to node " + nodeId + " which is not ready.");
-        } else if (!canSendRequest(nodeId))
-            throw new IllegalStateException("Attempt to send a request to node " + nodeId + " which is not ready.");
-
-        Send send = request.body().toSend(nodeId, request.header());
+    private void doSend(ClientRequest clientRequest, boolean isInternalRequest, long now) {
+        String nodeId = clientRequest.destination();
+        if (!isInternalRequest) {
+            // If this request came from outside the NetworkClient, validate
+            // that we can send data.  If the request is internal, we trust
+            // that that internal code has done this validation.  Validation
+            // will be slightly different for some internal requests (for
+            // example, ApiVersionsRequests can be sent prior to being in
+            // READY state.)
+            if (!canSendRequest(nodeId))
+                throw new IllegalStateException("Attempt to send a request to node " + nodeId + " which is not ready.");
+        }
+        AbstractRequest request = null;
+        AbstractRequest.Builder builder = clientRequest.requestBuilder();
```

Comment:
In the patch, it seems that once a channel is connected, but is not necessarily ready, NetworkClient can start issuing ApiRequest. This can fail if transport is SSL and transportLayer.ready() is false. It's also too late to only start issuing ApiRequest after channel is ready since the ApiRequest should be sent before SASL handshake (SASL handshake completes when authenticator.complete() is true). 

What the NetworkClient is supposed to do is only start issuing ApiRequest after transportLayer.ready() is true. It's just that currently, this state is no propagated by Selector.